### PR TITLE
More systematic checks that an estimator was fit before using its parameters

### PR DESCRIPTION
**Pull request by RobinVogel**
Solves #255

### CODE REVIEW

**Comment by terrytangyuan** [comment]
@RobinVogel It seems like this PR might have caused master branch to fail. Could you take a look? https://travis-ci.org/scikit-learn-contrib/metric-learn/builds/620639370

**Comment by RobinVogel** [comment]
Looking into it. The tests fail only on python 3 with skggm.
I had to install lapack before I could install skggm with pip with `conda install -c conda-forge lapack`,
followed by `pip install skggm`. For now I can't replicate the error of the travis log locally.

**Comment by bellet** [comment]
It is strange because all tests passed on merge:
https://travis-ci.org/scikit-learn-contrib/metric-learn/builds/618537553?utm_source=github_status&utm_medium=notification

File: test/test_pairs_classifiers.py [hunk 1]
```diff
@@ -133,10 +133,25 @@ def fit(self, pairs, y):
     pairs, y = self._prepare_inputs(pairs, y,
                                     type_of_inputs='tuples')
     self.components_ = np.atleast_2d(np.identity(pairs.shape[2]))
-    self.threshold_ = 'I am not set.'
+    # self.threshold_ is not set.
     return self
 
 
+def test_unset_threshold():
+  # test that set_threshold indeed sets the threshold
```
Comment (bellet) [inline test/test_pairs_classifiers.py]: this description is wrong

File: test/test_pairs_classifiers.py [hunk 2]
```diff
@@ -73,7 +73,7 @@ def test_raise_not_fitted_error_if_not_fitted(estimator, build_dataset,
   estimator.set_params(preprocessor=preprocessor)
   set_random_state(estimator)
   with pytest.raises(NotFittedError):
-    estimator.predict(input_data)
+    estimator.decision_function(input_data)
```
Comment (bellet) [inline test/test_pairs_classifiers.py]: a few things here:
- description of the test still mentions "predict"
- maybe all relevant methods should be tested here? it would have detected that there was a problem for `score_pairs`

**Comment by bellet** [comment]
I have re-run Travis to make sure it was not a glitch - but the errors are still there

File: test/test_pairs_classifiers.py (hunk 2 above)
Comment (RobinVogel) [reply test/test_pairs_classifiers.py]: Addressed in PR #269, sorry for the mishaps.

File: test/test_pairs_classifiers.py (hunk 1 above)
Comment (RobinVogel) [reply test/test_pairs_classifiers.py]: Addressed in PR #269, sorry for the mishap.

**Comment by bellet** [comment]
Fixing the failing tests in #270 
The reason the tests passed in the PR but failed on master after merging is because the new version of sklearn was released in between the two! Pretty big coincidence
